### ISSUE TITLE
Proper Dataset Formatting

### ISSUE THREAD

**Issue by SIN3R6Y** [lukalabs/cakechat#34]
I'm working on some some tools to automate the creation of my own training datasets. I've noticed that there are discrepancies between the supplied dummy data and the downloaded sets from AWS.

The dummy datasets include punctuation which till get tokenized as separate tokens when generating indices. But the AWS dataset does not include punctuation in the token index. Is the right plan of action to strip punctuation from the dataset?

Also there are placeholders like _unk_ and _pad_ that get added in as well, these are also not present in the AWS token index, but will be added to a generated index presumably do the default's in config.py

**Comment by SIN3R6Y** [comment]
Ah i see, so when building on top of a AWS dataset you want to keep the original index's. That does make more sense. Is there any reason to even run "prepare_index_files.py", assuming you are using the same conditions?

**Comment by nicolas-ivanov** [comment]
It makes sense to run `tools/prepare_index_files.py` if you are training your model from scratch on your own dataset that has a significantly different vocabulary – for example, your dataset is in French instead of English. 

In other cases it is reasonable to keep the provided `t_idx_processed_dialogs.json`, especially if you are using our pretrained model, since it is tuned to these tokens.

**Comment by SIN3R6Y** [comment]
What about in the case of your second dataset containing tokens that are not preset in the original token index? Wouldn't the right case then be to append them?

**Comment by nicolas-ivanov** [comment]
> i assume that there are restrictions which prevent distributing the original twitter corpus as it is likely obtained via paid access?

Yes, unfortunately we can't distribute our twitter corpus due to privacy restrictions, but you can collect one yourself: https://github.com/aritter/twitter_download
Or use one of many others publicly available corpora: https://breakend.github.io/DialogDatasets/

> What about in the case of your second dataset containing tokens that are not present in the original token index? Wouldn't the right case then be to append them?

There is no straightforward way to "add" new tokens to the trained model. The easiest approach is the following:

* Manually replace some rare tokens with your new words in `t_idx_processed_dialogs.json`
* Tune the trained model on your data that includes your new words
* Hope that you won't overfit! (toss up to you dataset some English-language samples to regularize the training process).

The embeddings for the new words will be screwed indeed, but they should fit the new tokens meaning quit quickly.

**Comment by SIN3R6Y** [comment]
> The embeddings for the new words will be screwed indeed, but they should fit the new tokens meaning quit quickly.

well from my brief testing i've found that it doesn't actually retrain the w2v model, even if you replace the tokens it still simply loads the old w2v model instead of appending the missing tokens to it. Even if it did append to the w2v model i'm not entirely sure you could accurately vectorize them without reprocessing the base corpus.

In reality it's probably best to have a full corpus available.

**Comment by nicolas-ivanov** [comment]
> even if you replace the tokens it still simply loads the old w2v model instead of appending the missing tokens to it

That's right: the script initialises token embedding layers with the values from the trained w2v model (if you set `USE_PRETRAINED_W2V_EMBEDDINGS_LAYER = True` in config) or with the values from the trained dialog model (if you set `USE_PRETRAINED_W2V_EMBEDDINGS_LAYER = False` in config – this is a preferable mode for tuning already trained model).

> Even if it did append to the w2v model i'm not entirely sure you could accurately vectorize them without reprocessing the base corpus.

1. With this approach you don't change w2v model anyhow, only the embedding layers of you dialog model.
2. The model should train well even without the help of pre-trained w3v embeddings initialisation and it should "accurately vectorize" the new tokens given enough time and training data.

> In reality it's probably best to have your own full corpus available.

Yep, that's the most reliable option.
